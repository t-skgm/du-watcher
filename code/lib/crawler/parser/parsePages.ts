import { parsePriceStr } from '@/utils/formatPrice'
import { HTMLElement, parse } from 'node-html-parser'

export const parsePages = (htmls: string[], baseUrl: string): UsedItem[] => {
  const results: UsedItem[] = []

  for (const html of htmls) {
    const doc = parse(html)
    const subGenreList = doc.querySelectorAll('.subGenreResult > .subGenreResult__li')

    if (subGenreList.length > 0) {
      results.push(...parseSubGenreResults(subGenreList, baseUrl))
    } else {
      const searchList = doc.querySelectorAll('.subGenreResult > .searchAll__li')
      results.push(...parseSearchResults(searchList, baseUrl))
    }
  }

  return results
}

const parseSubGenreResults = (resultList: HTMLElement[], baseUrl: string) => {
  const results: UsedItem[] = []

  for (const item of resultList) {
    const artist = item.querySelector('h3.subGenreResult__artist a')?.textContent?.trim()
    const productTitle = item.querySelector('h2.subGenreResult__name a')?.textContent?.trim()
    const labelName = item.querySelector('p.subGenreResult__other a')?.textContent?.trim()

    const isDiscountedPrice = item.querySelector('.u-priceDiscount') != null
    const discountRatePercentage = formatters.pickDiscountedRate(
      item.querySelector('.u-discountRate')?.textContent?.trim()
    )

    const cheapestItemPrice = isDiscountedPrice
      ? formatters.cutOffText(item.querySelector('p.u-priceDiscount')?.textContent?.trim())
      : item.querySelector('p.u-priceNormal--Blue')?.textContent?.trim()

    const cheapestItemStatus = item.querySelector('figure.subGenreResult__thumb > span > span')?.textContent?.trim()

    const genre = item.querySelector('p.subGenreResult__tag')?.textContent?.trim()
    const itemPageUrl = item.querySelector('h2.subGenreResult__name a')?.getAttribute('href')
    const itemId = ItemIdRegexp.exec(itemPageUrl ?? '')?.[1]
    const media = formatters.pickMediaText(item.querySelector('.subGenreResult__other')?.textContent ?? '')

    results.push({
      itemId,
      artist,
      productTitle,
      labelName,
      genre,
      cheapestItemPrice,
      cheapestItemStatus,
      media,
      isDiscountedPrice,
      discountRatePercentage,
      itemPageUrl: itemPageUrl ? `${baseUrl}${itemPageUrl}` : undefined,
      crawledAt: new Date()
    })
  }

  return results
}

const parseSearchResults = (resultList: HTMLElement[], baseUrl: string) => {
  const results: UsedItem[] = []

  for (const item of resultList) {
    const artist = item.querySelector('h2.searchAll__artist a')?.textContent?.trim()
    const productTitle = item.querySelector('h2.searchAll__name a')?.textContent?.replace(/\s+/g, ' ')?.trim()
    const labelName = item.querySelector('p.searchAll__other a')?.textContent?.trim()

    const isDiscountedPrice = item.querySelector('.u-priceDiscount') != null
    const discountRatePercentage = formatters.pickDiscountedRate(
      item.querySelector('.u-discountRate')?.textContent?.trim()
    )
    const normalItemPrice = isDiscountedPrice
      ? getPriceString(item.querySelector('p.u-priceDiscount'))
      : getPriceString(item.querySelector('p.u-priceNormal--Blue'))
    const normalItemStatus = item.querySelector('.tag-menbersSale')?.textContent?.trim()

    const subGenreUsedPrice = getPriceString(item.querySelector('.priceUsed__price'))
    // 最初のStatus
    const subGenreUsedStatus = item.querySelector('.qualityArea__li')?.textContent?.trim()

    const isNormalItemCheapest = normalItemPrice < subGenreUsedPrice
    const cheapestItemPrice = formatters.numberToPrice(isNormalItemCheapest ? normalItemPrice : subGenreUsedPrice)
    const cheapestItemStatus = isNormalItemCheapest ? normalItemStatus : subGenreUsedStatus

    const genre = item.querySelector('p.searchAll__tag')?.textContent?.trim()
    const itemPageUrl = item.querySelector('h2.searchAll__name a')?.getAttribute('href')
    const itemId = ItemIdRegexp.exec(itemPageUrl ?? '')?.[1]
    const media = formatters.pickMediaText(item.querySelector('.searchAll__other')?.textContent ?? '')

    results.push({
      itemId,
      artist,
      productTitle,
      labelName,
      genre,
      cheapestItemPrice,
      cheapestItemStatus,
      media,
      isDiscountedPrice,
      discountRatePercentage,
      itemPageUrl: itemPageUrl ? `${baseUrl}${itemPageUrl}` : undefined,
      crawledAt: new Date()
    })
  }

  return results
}

export type UsedItem = {
  artist: string | undefined
  productTitle: string | undefined
  labelName: string | undefined
  genre: string | undefined
  cheapestItemPrice: string | undefined
  cheapestItemStatus: string | undefined
  isDiscountedPrice: boolean
  discountRatePercentage: string | undefined
  media: string | undefined
  itemPageUrl: string | undefined
  itemId: string | undefined
  crawledAt: Date
}

const ItemIdRegexp = /detail\/(.*)/

const formatters = {
  // X%OFF -> X
  pickDiscountedRate: (text: string | undefined) => text?.replace(/(\d+)%OFF$/, '$1'),
  cutOffText: (text: string | undefined) => text?.replace(/^\d+%OFF\s*/, ''),
  pickMediaText: (text: string | undefined) => text?.replace(/\s+/g, ' ')?.split(' / ')[2].trim(),
  // ex. 1000 -> 1,000円
  numberToPrice: (priceNum: number) => `${priceNum.toLocaleString()}円(税込)`
}

const getPriceString = (priceNode: HTMLElement | null): number => {
  if (!priceNode) return Number.MAX_SAFE_INTEGER
  priceNode.querySelectorAll('span').forEach(span => span.remove())
  return parsePriceStr(priceNode.textContent.trim()) ?? Number.MAX_SAFE_INTEGER
}
