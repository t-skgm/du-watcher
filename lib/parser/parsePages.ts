import { parse } from 'node-html-parser'
import { DU_BASE_URL } from './const'

export const parsePages = (htmls: string[]): UsedItem[] => {
  const results: UsedItem[] = []

  htmls.forEach(html => {
    const doc = parse(html)
    const resultList = doc.querySelectorAll('.subGenreResult > .subGenreResult__li')

    resultList.forEach(item => {
      const artist = item.querySelector('h3.subGenreResult__artist a')?.textContent.trim()
      const productTitle = item.querySelector('h2.subGenreResult__name a')?.textContent.trim()
      const labelName = item.querySelector('p.subGenreResult__other a')?.textContent.trim()
      const cheapestItemPrice = item.querySelector('p.u-priceNormal--Blue')?.textContent.trim()
      const cheapestItemStatus = item.querySelector('figure.subGenreResult__thumb > span > span')?.textContent.trim()
      const genre = item.querySelector('p.subGenreResult__tag')?.textContent.trim()
      const itemPageUrl = item.querySelector('h2.subGenreResult__name a')?.getAttribute('href')
      const itemId = ItemIdRegexp.exec(itemPageUrl ?? '')?.[1]

      results.push({
        itemId,
        artist,
        productTitle,
        labelName,
        genre,
        cheapestItemPrice,
        cheapestItemStatus,
        itemPageUrl: itemPageUrl && `${DU_BASE_URL}${itemPageUrl}`,
        crawledAt: new Date()
      })
    })
  })

  return results
}

export type UsedItem = {
  artist: string | undefined
  productTitle: string | undefined
  labelName: string | undefined
  genre: string | undefined
  cheapestItemPrice: string | undefined
  cheapestItemStatus: string | undefined
  itemPageUrl: string | undefined
  itemId: string | undefined
  crawledAt: Date
}

const ItemIdRegexp = /udetail\/(.*)/
