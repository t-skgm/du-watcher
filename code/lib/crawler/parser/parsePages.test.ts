import { afterEach, beforeEach, describe, expect, test, setSystemTime } from 'bun:test'
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { parsePages } from './parsePages'

describe('parsePages()', () => {
  beforeEach(() => {
    const date = new Date('2112-07-03T00:00:00+09:00')
    setSystemTime(date)
  })

  afterEach(() => {
    setSystemTime()
  })

  describe('genre: normal', () => {
    const snapshotHtml = readFileSync(resolve(__dirname, './samples/list-genre-normal.snapshot.html'), {
      encoding: 'utf-8'
    })
    test('件数', async () => {
      const result = parsePages([snapshotHtml], 'https://base-url.exmaple')

      expect(result).toHaveLength(50)
    })

    test('最初の3件', async () => {
      const result = parsePages([snapshotHtml], 'https://base-url.exmaple')
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toMatchSnapshot()
      expect(result[2]).toMatchSnapshot()
    })
  })

  describe('genre: discounted', () => {
    const snapshotHtml = readFileSync(resolve(__dirname, './samples/list-genre-discount.snapshot.html'), {
      encoding: 'utf-8'
    })
    test('件数', async () => {
      const result = parsePages([snapshotHtml], 'https://base-url.exmaple')

      expect(result).toHaveLength(50)
    })

    test('最初の3件', async () => {
      const result = parsePages([snapshotHtml], 'https://base-url.exmaple')
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toMatchSnapshot()
      expect(result[2]).toMatchSnapshot()
    })
  })

  describe('all: discounted', () => {
    const snapshotHtml = readFileSync(resolve(__dirname, './samples/list-all-discount.snapshot.html'), {
      encoding: 'utf-8'
    })
    test('件数', async () => {
      const result = parsePages([snapshotHtml], 'https://base-url.exmaple')

      expect(result).toHaveLength(50)
    })

    test('最初の3件', async () => {
      const result = parsePages([snapshotHtml], 'https://base-url.exmaple')
      expect(result[0]).toMatchSnapshot()
      expect(result[1]).toMatchSnapshot()
      expect(result[2]).toMatchSnapshot()
    })
  })
})
