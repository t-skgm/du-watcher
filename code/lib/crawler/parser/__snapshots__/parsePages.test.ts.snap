// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePages() genre: normal 最初の3件 1`] = `
{
  "artist": "BEN WATT",
  "cheapestItemPrice": "2,250円(税込)",
  "cheapestItemStatus": "B",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": undefined,
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": false,
  "itemId": "1007488350",
  "itemPageUrl": "https://base-url.exmaple/used/ct/indiealt/udetail/1007488350",
  "labelName": "CHERRY RED",
  "media": "12"(レコード)",
  "productTitle": "SUMMER INTO WINTER",
}
`;

exports[`parsePages() genre: normal 最初の3件 2`] = `
{
  "artist": "2ND GRADE",
  "cheapestItemPrice": "2,035円(税込)",
  "cheapestItemStatus": "S",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "50",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008552801",
  "itemPageUrl": "https://base-url.exmaple/used/ct/indiealt/udetail/1008552801",
  "labelName": "DOUBLE DOUBLE WHAMMY",
  "media": "LP(レコード)",
  "productTitle": "EASY LISTENING (VINYL)",
}
`;

exports[`parsePages() genre: normal 最初の3件 3`] = `
{
  "artist": "ALEX G アレックス・G",
  "cheapestItemPrice": "2,695円(税込)",
  "cheapestItemStatus": "B",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "30",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008448920",
  "itemPageUrl": "https://base-url.exmaple/used/ct/indiealt/udetail/1008448920",
  "labelName": "LUCKY NUMBER",
  "media": "LP(レコード)",
  "productTitle": "RULES (LP+7")",
}
`;

exports[`parsePages() genre: discounted 最初の3件 1`] = `
{
  "artist": "A'S (INDIE)",
  "cheapestItemPrice": "1,709円(税込)",
  "cheapestItemStatus": "B",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "5",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008509505",
  "itemPageUrl": "https://base-url.exmaple/used/ct/indiealt/udetail/1008509505",
  "labelName": "PSYCHIC HOTLINE",
  "media": "LP(レコード)",
  "productTitle": "FRUIT (BLACK VINYL)",
}
`;

exports[`parsePages() genre: discounted 最初の3件 2`] = `
{
  "artist": "ALEX CALDER",
  "cheapestItemPrice": "1,045円(税込)",
  "cheapestItemStatus": "B",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "5",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "AWS150123-AC1",
  "itemPageUrl": "https://base-url.exmaple/used/ct/indiealt/udetail/AWS150123-AC1",
  "labelName": "CAPTURED TRACKS",
  "media": "LP(レコード)",
  "productTitle": "STRANGE DREAMS (LP)",
}
`;

exports[`parsePages() genre: discounted 最初の3件 3`] = `
{
  "artist": "ALEX G アレックス・G",
  "cheapestItemPrice": "2,232円(税込)",
  "cheapestItemStatus": "S",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "5",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008448920",
  "itemPageUrl": "https://base-url.exmaple/used/ct/indiealt/udetail/1008448920",
  "labelName": "LUCKY NUMBER",
  "media": "LP(レコード)",
  "productTitle": "RULES (LP+7")",
}
`;


exports[`parsePages() jrock 最初の3件 1`] = `
{
  "artist": "BLANKEY JET CITY ブランキー・ジェット・シティ",
  "cheapestItemPrice": "5,300円(税込)",
  "cheapestItemStatus": "A",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": undefined,
  "genre": "日本のロック",
  "isDiscountedPrice": false,
  "itemId": "1008930066",
  "itemPageUrl": "https://base-url.exmaple/jp/ct/detail/1008930066",
  "labelName": "ユニバーサルミュージック",
  "media": "LP(レコード)",
  "productTitle": "THE SIX",
}
`;

exports[`parsePages() jrock 最初の3件 2`] = `
{
  "artist": "TAKESHI INOMATA 猪俣猛",
  "cheapestItemPrice": "2,650円(税込)",
  "cheapestItemStatus": "A",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "30",
  "genre": "JAZZ",
  "isDiscountedPrice": true,
  "itemId": "1008861334",
  "itemPageUrl": "https://base-url.exmaple/jp/ct/detail/1008861334",
  "labelName": "ユニバーサルミュージック",
  "media": "LP(レコード)",
  "productTitle": "New Rock In Europe(LP)",
}
`;

exports[`parsePages() jrock 最初の3件 3`] = `
{
  "artist": "帯化",
  "cheapestItemPrice": "2,000円(税込)",
  "cheapestItemStatus": "OUTLET",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": undefined,
  "genre": "日本のロック",
  "isDiscountedPrice": true,
  "itemId": "1008900421",
  "itemPageUrl": "https://base-url.exmaple/jp/ct/detail/1008900421",
  "labelName": "造園計画",
  "media": "LP(レコード)",
  "productTitle": "御池塘自治 (LP)",
}
`;

exports[`parsePages() all: discounted 最初の3件 1`] = `
{
  "artist": "DUNGEN ドゥンエン",
  "cheapestItemPrice": "1,980円(税込)",
  "cheapestItemStatus": "OUTLET",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "60",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008525007",
  "itemPageUrl": "https://base-url.exmaple/indiealt/ct/detail/1008525007",
  "labelName": "MEXICAN SUMMER",
  "media": "LP(レコード)",
  "productTitle": "EN AR FOR MYCKET OCH TUSEN ALDRIG NOG(COLOR VINYL)",
}
`;

exports[`parsePages() all: discounted 最初の3件 2`] = `
{
  "artist": "KING KRULE キング・クルール",
  "cheapestItemPrice": "1,672円(税込)",
  "cheapestItemStatus": "OUTLET",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "60",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008365027",
  "itemPageUrl": "https://base-url.exmaple/indiealt/ct/detail/1008365027",
  "labelName": "XL RECORDINGS",
  "media": "LP(レコード)",
  "productTitle": "YOU HEAT ME UP, YOU COOL ME DOWN(LP)",
}
`;

exports[`parsePages() all: discounted 最初の3件 3`] = `
{
  "artist": "CHIME SCHOOL チャイム・スクール",
  "cheapestItemPrice": "1,496円(税込)",
  "cheapestItemStatus": "OUTLET",
  "crawledAt": 2112-07-02T15:00:00.000Z,
  "discountRatePercentage": "20",
  "genre": "ROCK / POPS / INDIE",
  "isDiscountedPrice": true,
  "itemId": "1008523216",
  "itemPageUrl": "https://base-url.exmaple/indiealt/ct/detail/1008523216",
  "labelName": "SLUMBERLAND",
  "media": "CASSETTE TAPE",
  "productTitle": "CHIME SCHOOL (CASSETTE TAPE)",
}
`;
